### PR TITLE
Conforms `Request.Body` to  `AsyncSequence`

### PR DESCRIPTION
**Pull request by mcritz**
This PR wraps `Request.Body.drain()` as a modern Swift `AsyncSequence<ByteBuffer, Error>`.

This is useful to _stream_ bytes from request rather than collecting them in memory. Example: A route could handle a multigigbyte file upload like this:

```swift
do {
    let nioFileHandle = try NIOFileHandle(path: filePath, mode: .write)
    var offset: Int64 = 0
    
    for try await bytes in req.body {
        try await req.application.fileio.write(fileHandle: nioFileHandle,
                                               toOffset: offset,
                                               buffer: bytes,
                                               eventLoop: req.eventLoop).get()
        offset += Int64(bytes.readableBytes)
        try nioFileHandle.close()
    }
} catch {
   ...
}
```

### CODE REVIEW

**Comment by mcritz** [comment]
Hello! I’m starting this Draft PR as a conversation starter for adding `AsyncThrowingStream` API to Vapor. If this work is considered worthwhile by the maintainers, then I can perform any changes and write tests.

Cheers

**Comment by 0xTim** [comment]
Just to carry over from Discord - this looks great! Will need some tests though

**Comment by mcritz** [comment]
I have refactored this PR to support backpressure with AsyncThrowingChannel.

Adds a dependency on `swift-async-algorithms`.

The call site API changes a little bit:

```
do {
    let nioFileHandle = try NIOFileHandle(path: filePath, mode: .write)
    var offset: Int64 = 0
    let channel = req.asyncThrowingChannel
    for try await byteBuffer in channel {
        try await req.application.fileio.write(fileHandle: nioFileHandle,
                                               toOffset: offset,
                                               buffer: byteBuffer,
                                               eventLoop: req.eventLoop).get()
        offset += Int64(byteBuffer.readableBytes)
    }
    try nioFileHandle.close()
} catch {
    ...
}
```

**Comment by mcritz** [comment]
Hi maintainers!

#2839 is ready for review. I am open to any changes.

In addition to the included test, I have built a [test project](https://github.com/mcritz/VaporUploads/tree/feature/async) that proves out the use case. I ran a Swift Concurrency Profile run against the test project target while uploading files and it met my expectations for concurrent tasks.

Cheers!

File: Package.swift [hunk 1]
```diff
@@ -52,7 +52,7 @@ let package = Package(
         
         // Swift collection algorithms
         .package(url: "https://github.com/apple/swift-algorithms.git", from: "1.0.0"),
-
+
```
Comment (Joannis) [inline Package.swift:55]: Could we get rid of this change? Just for cleanliness

File: Sources/Vapor/Concurrency/Request+Concurrency.swift [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+                let result = source.yield(buffer)
+                // Inspect the source view and handle outcomes
+                switch result {
+                case .dropped:
+                    // The consumer dropped the sequence.
+                    // Inform the producer that we don't want more data
+                    // by returning an error in the future.
+                    return request.eventLoop.makeFailedFuture(CancellationError())
+                case .stopProducing:
+                    // The consumer is consuming fast enough for us.
```
Comment (Joannis) [inline Sources/Vapor/Concurrency/Request+Concurrency.swift:147]: Isn't it the opposite, the consumer is **not** consuming fast enough?

**Review by Joannis — APPROVED** [review]
LGTM, just two nits

File: Package.swift (hunk 1 above)
Comment (mcritz) [reply Package.swift]: done

**Comment by mcritz** [comment]
@Joannis Updated! Thanks for review.

**Review by 0xTim — APPROVED** [review]
This looks great thanks!

**Comment by 0xTim** [comment]
@gwynne any final thoughts before merging

**Review by gwynne — APPROVED** [review]
I could nitpick a few lines over different ways to write exactly the same code that make no actual difference whatsoever... 😆 But nah, LGTM!

**Comment by gwynne** [comment]
Although the CI failures should be addressed as appropriate - I probably should've finished scrolling before commenting 😅

**Comment by 0xTim** [comment]
@mcritz we either need to gate this behind 5.7 only or rewrite to allow it to work on older versions of Swift

**Comment by mcritz** [comment]
@0xTim On it. 

@gwynne Please nitpick. I'd like to learn.

**Pull request merged**